src/appleseed-max-impl/appleseedrenderer/dialoglogtarget.cpp

```diff
@@ -0,0 +1,259 @@
+                    message_color = RGB(10, 150, 10);
+                    break;
+                  default:
+                    message_color = RGB(10, 10, 10);
+                    break;
+                }
+                append_text(GetDlgItem(g_log_dialog, IDC_EDIT_LOG), utf8_to_wide(line + "\n").c_str(), message_color);
+            }
+        }
+    }
+
+    void push_message(MessagePair message)
+    {
+        boost::mutex::scoped_lock lock(g_message_queue_mutex);
+        g_message_queue.push_back(message);
+    }
+
+    // Runs in UI thread.
+    void emit_pending_messages()
+    {
+        if (!g_log_dialog)
+        {
+            g_log_dialog = CreateDialogParam(
+                g_module,
+                MAKEINTRESOURCE(IDD_DIALOG_LOG),
+                GetCOREInterface()->GetMAXHWnd(),
+                dialog_proc,
+                NULL);
+
+            GetCOREInterface14()->RegisterModelessRenderWindow(g_log_dialog);
+        }
+
+        if (g_log_dialog != NULL)
+        {
+            boost::mutex::scoped_lock lock(g_message_queue_mutex);
+
+            for (const auto& message : g_message_queue)
+                print_message(message.first, message.second);
+
+            g_message_queue.clear();
+        }
+    }
+
+    // Runs in UI thread.
+    void emit_saved_messages()
+    {
+        if (g_log_dialog)
+            SetDlgItemText(g_log_dialog, IDC_EDIT_LOG, L"");
+
+        emit_pending_messages();
+    }
+
+    const UINT WM_TRIGGER_CALLBACK = WM_USER + 4764;
+}
+
+DialogLogTarget::DialogLogTarget(DialogLogMode open_mode)
+  : m_log_mode(open_mode)
+{
+    m_session_messages.clear();
+    g_message_queue.clear();
+    asr::global_logger().add_target(this);
+}
+
+void DialogLogTarget::release()
+{
+    asr::global_logger().remove_target(this);
+    delete this;
+}
+
+void DialogLogTarget::write(
+    const MessageType       category,
+    const char*             file,
+    const size_t            line,
+    const char*             header,
+    const char*             message)
+{
+    std::vector<std::string> lines;
+    asf::split(message, "\n", lines);
+
+    m_session_messages.push_back(MessagePair(category, lines));
+
+    push_message(MessagePair(category, lines));
+
+    if (g_log_dialog)
+        print_to_dialog();
+    else
+    {
+        switch (category)
+        {
+          case asf::LogMessage::Category::Error:
+          case asf::LogMessage::Category::Fatal:
+          case asf::LogMessage::Category::Warning:
+            if (m_log_mode != DialogLogMode::Never)
+                print_to_dialog();
+            break;
+          case asf::LogMessage::Category::Debug:
+          case asf::LogMessage::Category::Info:
+            if (m_log_mode == DialogLogMode::Always)
+                print_to_dialog();
+            break;
+        }
+    }
+}
+
+void DialogLogTarget::show_last_session_messages()
+{
+    if (g_log_dialog)
+        return;
+
+    boost::mutex::scoped_lock lock(g_message_queue_mutex);
```

Comment:
Wrap this block in curly braces such that the mutex gets released before the call to `PostMessage()`.